### PR TITLE
envs robotics - touch sensors - Shadow Hand

### PR DESCRIPTION
**Pull request by ndrwmlnk**
@matthiasplappert
We (@ndrwmlnk @llach) covered the hand model with 92 touch sensors and extended the HandManipulate{Block, Egg, Pen}-v0 environments to HandManipulate{Block, Egg, Pen}TouchSensors-v0. Now, an observation additionally contains 92 touch values (0 - no touch, 1 - touch detected). Sensory information allows faster and better convergence (preliminary results).

Test the environments with pre-trained weights (currently in training):
https://rebrand.ly/TouchSensors
python baselines/her/experiment/play.py HandManipulateBlockTouchSensors.pkl
python baselines/her/experiment/play.py HandManipulateEggTouchSensors.pkl
python baselines/her/experiment/play.py HandManipulatePenTouchSensors.pkl

Touch sensor sites are tailored to physical geoms (robot0:DC_Hand), thus may visually overlap with meshes (robot0:D_Vizual). Active touch sensors are highlighted in red, inactive sensors - in green:
https://rebrand.ly/TouchSensors
sensors_cube_geoms.mp4
sensors_cube_mesh.mp4
sensors_cube.mp4
sensors.mp4

### CODE REVIEW

**Comment by matthiasplappert** [comment]
Thanks for your PR! I'm excited to get this merged and will work with @pzhokhov to figure out the best next steps.

File: gym/envs/robotics/assets/hand/shared.xml [hunk 1]
```diff
@@ -19,9 +19,13 @@
         </default>
         <default class="robot0:DC_Hand">
             <geom material="robot0:MatColl" contype="1" conaffinity="0" group="4"></geom>
+            <!--visualize physical bodies, hide ghost mesh-->
```
Comment (matthiasplappert) [inline gym/envs/robotics/assets/hand/shared.xml]: Why was this change necessary?

File: gym/envs/robotics/hand/manipulate_touch_sensors.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            touch_get_obs (string): touch sensor readings
+                - "boolean": returns 1 if touch sensor reading != 0.0 else 0
+                - "log": returns log(x+1) touch sensor readings from self.sim.data.sensordata[]
+                - "off" or else: does not add touch sensor readings to the observation
+
+        """
+        self.touch_visualisation = touch_visualisation
+        self.touch_get_obs = touch_get_obs
+        self._tsensor_prefix = 'robot0:TS_'
+        # self._tsensor_group_names = []  # list of dictionaries of sensors per joint
```
Comment (matthiasplappert) [inline gym/envs/robotics/hand/manipulate_touch_sensors.py]: Please remove these commented-out lines.

File: gym/envs/__init__.py [hunk 3]
```diff
@@ -377,6 +377,13 @@ def _merge(a, b):
         max_episode_steps=100,
     )
 
+    register(
```
Comment (matthiasplappert) [inline gym/envs/__init__.py]: Could you please also register the environments for `Egg` and `Pen` with touch sensing here?

File: gym/envs/robotics/hand/manipulate_touch_sensors.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    axis /= np.linalg.norm(axis)
+    quat = np.concatenate([[np.cos(angle / 2.)], np.sin(angle / 2.) * axis])
+    quat /= np.linalg.norm(quat)
+    return quat
+
+
+# Ensure we get the path separator correct on windows
+MANIPULATE_BLOCK_XML = os.path.join('hand', 'manipulate_block_touch_sensors_76.xml')
+MANIPULATE_EGG_XML = os.path.join('hand', 'manipulate_egg.xml')
+MANIPULATE_PEN_XML = os.path.join('hand', 'manipulate_pen.xml')
```
Comment (matthiasplappert) [inline gym/envs/robotics/hand/manipulate_touch_sensors.py]: Would be nice to have touch sensor support here as well. If not, please remove them, since it is otherwise confusing.

File: gym/envs/robotics/hand/manipulate_touch_sensors.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+from gym.envs.robotics import rotations, hand_env
+from gym.envs.robotics.hand import manipulate
+
+try:
+    import mujoco_py
+except ImportError as e:
+    raise error.DependencyNotInstalled("{}. (HINT: you need to install mujoco_py, and also perform the setup instructions here: https://github.com/openai/mujoco-py/.)".format(e))
+
+
+def quat_from_angle_and_axis(angle, axis):
```
Comment (matthiasplappert) [inline gym/envs/robotics/hand/manipulate_touch_sensors.py]: Please re-use the implementation from `manipulate.py` for this function

File: gym/envs/robotics/hand/manipulate_touch_sensors.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            target_position_range, reward_type, initial_qpos=initial_qpos,
+            randomize_initial_position=randomize_initial_position, randomize_initial_rotation=randomize_initial_rotation,
+            distance_threshold=distance_threshold, rotation_threshold=rotation_threshold, n_substeps=n_substeps, relative_control=relative_control,
+            ignore_z_target_rotation=ignore_z_target_rotation,
+        )
+        utils.EzPickle.__init__(self)
+
+        for k, v in self.sim.model._sensor_id2name.items():  # get touch sensor ids and their site names
+            if 'TS' in v:
+                # self._tsensor_name2id[v] = k
```
Comment (matthiasplappert) [inline gym/envs/robotics/hand/manipulate_touch_sensors.py]: Please remove this line.

File: gym/envs/robotics/hand/manipulate_touch_sensors.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            for k, v in self._tsensor_id2name.items():
+                self.sim.model.site_rgba[self._tsensor_id2siteid[k]] = [0, 0, 0, 0]
+
+        return {
+            'observation': observation.copy(),
+            'achieved_goal': achieved_goal.copy(),
+            'desired_goal': self.goal.ravel().copy(),
+        }
+
+    def set_object_size(self, factor):  # changes size of the manipulated object
```
Comment (matthiasplappert) [inline gym/envs/robotics/hand/manipulate_touch_sensors.py]: Please move this implementation to the parent class since it's equally applicable to the non-touch version.

File: gym/envs/robotics/hand/manipulate_touch_sensors.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                touch_values.append(value)
+        elif self.touch_get_obs == 'log':
+            for k, v in self._tsensor_id2name.items():
+                value = self.sim.data.sensordata[k]
+                touch_values.append(value)
+            if len(touch_values) > 0:
+                touch_values = np.log(np.array(touch_values) + 1.0)
+        observation = np.concatenate([robot_qpos, robot_qvel, object_qvel, touch_values, achieved_goal])
+
+        # set rgba values
```
Comment (matthiasplappert) [inline gym/envs/robotics/hand/manipulate_touch_sensors.py]: Please move everything related to visualization to `_render_callback` (https://github.com/openai/gym/blob/master/gym/envs/robotics/robot_env.py#L152) instead. Don't forget to call `super` since the hand environment itself does some visualization here as well.

File: gym/envs/robotics/hand/manipulate_touch_sensors.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        if self.touch_visualisation == 'always':
+            for k, v in self._tsensor_id2name.items():
+                self.sim.model.site_rgba[self._tsensor_id2siteid[k]] = self._site_id2intial_rgba[self._tsensor_id2siteid[k]].copy()
+        elif self.touch_visualisation == 'on_touch':
+            for k, v in self._tsensor_id2name.items():
+                if self.sim.data.sensordata[k] != 0.0:
+                    # self.sim.model.site_rgba[self._tsensor_id2siteid[k]] = self._site_id2intial_rgba[self._tsensor_id2siteid[k]].copy()
```
Comment (matthiasplappert) [inline gym/envs/robotics/hand/manipulate_touch_sensors.py]: Please remove this line.

**Comment by pzhokhov** [comment]
my 2 cents here are the following - a couple of tests of the new environments are failing (tried running manually, unfortunately, those won't show in travis because this is an external PR and as such our mujoco key is not injected => no tests of mujoco-based envs are run). Please run them manually (via `pytest .`  with MUJOCO_KEY env variable set up and make sure 1) tests pass and 2) test count is 359 (that's how much it counted for me when I tried cloning the PR branch and running tests there). If there are some failures that seem unrelated to your changes, let's work together and figure out what's wrong

**Comment by ndrwmlnk** [comment]
@matthiasplappert @pzhokhov
Thanks for reviewing the code and tips with pytest .

I improved the code and fixed requested changes. 376 tests passed without errors. The number of tests exceeds 359, probably because of the new touch-sensory robotic environments.
The reason for the previous pytest error was the length of the observation vector:
https://github.com/openai/gym/blob/49cd48020f6760630a7317cb3529a22de6f12f2e/gym/envs/tests/test_envs.py#L23

I added obs to ManipulateTouchSensorsEnv.__init__ here:
https://github.com/ndrwmlnk/gym/blob/0d71b27103cd434173eb24ca50a99b3968866cfd/gym/envs/robotics/hand/manipulate_touch_sensors.py#L59
which now overrides obs from RobotEnv.__init__ here (without touch values):
https://github.com/openai/gym/blob/49cd48020f6760630a7317cb3529a22de6f12f2e/gym/envs/robotics/robot_env.py#L38

Please let me know if I can make any further improvements.

**Comment by pzhokhov** [comment]
@matthiasplappert @ndrwmlnk confirmed - tests are passing; if @matthiasplappert is happy, I am happy :)

**Comment by pzhokhov** [comment]
here's a successful travis build with the changes: https://travis-ci.org/openai/gym/jobs/489766042 (pay the "All checks have failed" no heed, I messed something up with github webhook when trying to set a build with the changes up)

**Comment by matthiasplappert** [comment]
Will take another pass tomorrow (crazy busy week so far) and then we can hopefully get this merged. Thanks everyone for helping!

**Comment by matthiasplappert** [comment]
And thanks for working with us on this PR; happy that we got it merged :)
